### PR TITLE
added: OC: Allow user to open next form when closing current form

### PR DESCRIPTION
**Pull request by theywa**
#320

### CODE REVIEW

File: app/views/styles/component/_form_footer.scss [hunk 1]
```diff
@@ -145,6 +145,62 @@
         &__main-controls {
             margin-bottom: 0;
             margin-top: 0;
+
+            .next-prompt-container {
+                width: 100%;
+                margin-bottom: 10px;
+                text-align: center;
+
+                input[type=checkbox] {
```
Comment (MartijnR) [inline app/views/styles/component/_form_footer.scss:154]: To get the same checkbox styling as in the form, you can simply change the HTML syntax a bit. See the removed code in _form-footer.pug in this commit: https://github.com/enketo/enketo-express/commit/15b29116023123acf8d4781b5f3d2d1298ed7340 (we should have a tickbox for save as draft). I believe the `<label class="question"><input type="checkbox"><span class="option-label">some text</span></label>` structure will do the trick. CSS rules are here (in enketo-core): https://github.com/enketo/enketo-core/blob/dbc18f2cb382e58af98c90138b443ac2407dd61a/src/sass/formhub/_radiocheck.scss#L26

File: app/views/surveys/component/_form-footer.pug [hunk 2]
```diff
@@ -7,6 +7,10 @@ section.form-footer
             i.icon.icon-check= ' '
             span(data-i18n='formfooter.validate.btn')= t('formfooter.validate.btn')
         else if type === 'fieldsubmission' || type === 'view'
```
Comment (MartijnR) [inline app/views/surveys/component/_form-footer.pug:9]: nextPrompt may require an additional check to make sure it is not added in views that should not support it (even if protection has been added to Enketo's OC API). Tbc, depends on Paul's answer.

**Review by MartijnR — CHANGES_REQUESTED** [review]
Looks great. Just a few simplification comments.

Also, please add some oc-api tests for endpoints that support this and for those that don't.

File: public/js/src/module/controller-webform.js [hunk 3]
```diff
@@ -686,10 +686,12 @@ function inIframe() {
  * @param  {Event} event - event
  */
 function postEventAsMessageToParentWindow( event ) {
+    const nextPrompt = document.querySelector( 'input[name=next-prompt]' );
     if ( event && event.type ) {
         try {
             window.parent.postMessage( JSON.stringify( {
-                enketoEvent: event.type
+                enketoEvent: event.type,
+                nextForm: nextPrompt && nextPrompt.checked ? true : false
```
[the remark this answers is not in the document]
Comment (theywa) [reply public/js/src/module/controller-webform.js]: yes you are right, code updated

File: app/views/styles/component/_form_footer.scss (hunk 1 above)
Comment (theywa) [reply app/views/styles/component/_form_footer.scss:152]: got it, code updated

**Comment by theywa** [comment]
oc-api-test updated

File: app/views/surveys/component/_form-footer.pug (hunk 2 above)
Comment (theywa) [reply app/views/surveys/component/_form-footer.pug:14]: updated, support on preview too

**Comment by theywa** [comment]
@MartijnR yes please, I just update the code to enabled on preview too

File: app/views/styles/component/_form_footer.scss [hunk 4]
```diff
@@ -145,6 +145,18 @@
...
+                    float: none;
+                }
+
+                span {
+                    vertical-align: top;
+                }
+            }
         }
     }
```
Comment (MartijnR) [inline app/views/styles/component/_form_footer.scss:162]: There is an issue with non-grid themes (eg. the OC theme):

Screen Shot 2020-11-16 at 3 39 50 PM

Comment (MartijnR) [reply app/views/styles/component/_form_footer.scss:162]: In the Grid theme, would be good to tweak the hover behavior. Doesn't look so good.

Screen Shot 2020-11-16 at 3 12 54 PM

At least override the yellow background (to transparent) I think

File: app/views/surveys/component/_form-footer.pug [hunk 5]
```diff
@@ -3,10 +3,20 @@ section.form-footer
     if !headless
       .form-footer__content__main-controls  
         if type === 'preview'
+          if nextPrompt
+            fieldset.draft.question.next-prompt
```
Comment (MartijnR) [inline app/views/surveys/component/_form-footer.pug:7]: The "draft" class doesn't do anything I believe and is confusing. You can target this question with `.next-prompt` in CSS

File: app/views/surveys/component/_form-footer.pug [hunk 6]
```diff
@@ -3,10 +3,20 @@ section.form-footer
...
+                span #{nextPrompt}
           button#validate-form.btn.btn-primary
             i.icon.icon-check= ' '
             span(data-i18n='formfooter.validate.btn')= t('formfooter.validate.btn')
         else if type === 'fieldsubmission' || type === 'view'
+          if nextPrompt && completeButton
+            fieldset.draft.question.next-prompt
+              .option-wrapper: label 
+                input.ignore(type='checkbox', name='next-prompt') 
+                span #{nextPrompt}
```
Comment (MartijnR) [inline app/views/surveys/component/_form-footer.pug:19]: might be good to give the span the 'option-label' class (like in the form) so the styling moves along with any changes to checkbox form styling (in the future). However, if you're engaged in an annoying CSS fight when doing this, don't bother.

**Review by MartijnR — CHANGES_REQUESTED** [review]

File: app/views/styles/component/_form_footer.scss (hunk 4 above)
Comment (theywa) [reply app/views/styles/component/_form_footer.scss:172]: style updated
using oc theme
Screenshot from 2020-11-18 15-01-30
using grid theme
Screenshot from 2020-11-18 15-05-55
is it okay to reduce the length of the background color when hovering?

File: app/views/surveys/component/_form-footer.pug (hunk 5 above)
Comment (theywa) [reply app/views/surveys/component/_form-footer.pug]: got it, style updated

File: app/views/surveys/component/_form-footer.pug (hunk 6 above)
Comment (theywa) [reply app/views/surveys/component/_form-footer.pug]: sure style updated

File: app/views/styles/component/_form_footer.scss (hunk 4 above)
Comment (MartijnR) [reply app/views/styles/component/_form_footer.scss:172]: Yes, I think that's great.

**Comment by MartijnR** [comment]
Thanks @theywa,

Sorry for my inefficient review process. I have to work on my review skills...

I found one remaining issue for single-page forms. I think this link may work for you: http://localhost:8005/single/fs/i/widgets?ecid=a&next_prompt=how%20are%20you%3F

Screen Shot 2020-11-20 at 11 27 02 AM

I also overlooked something I should have commented on before (though easy to change). Try to keep CSS changes limited to custom OC files. In this case the _form_footer.scss changes could be moved to _common-oc.scss. You'll find some other form_footer customizations in that file already. The reason for doing that is to reduce merge conflicts with enketo/enketo-express.

**Comment by theywa** [comment]
@MartijnR ~ No it's okay, I should check it more before making a PR
Update on single page(http://localhost:8888/single/fs/i/9ogdxV7K?ecid=a&parent_window_origin=https%3A%2F%2Fwww.w3schools.com&next_prompt=Test%20CheckBox) and move style code to _common-oc.scss
single-page

File: app/views/styles/component/_form_footer.scss [hunk 7]
```diff
@@ -147,12 +147,6 @@
             margin-top: 0;
         }
     }
-
-    &.end {
-        .logout {
-            margin-bottom: 50px;
-        }
-    }
 }
```
Comment (MartijnR) [inline app/views/styles/component/_form_footer.scss:150]: I believe this got accidentally deleted.

**Review by MartijnR — CHANGES_REQUESTED** [review]

File: app/views/styles/component/_form_footer.scss (hunk 7 above)
Comment (theywa) [reply app/views/styles/component/_form_footer.scss]: yes, sorry about that. it's clean now

**Pull request merged**